### PR TITLE
[NOD-1416] Implement BlockProcessor.

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: domain/consensus/processes/blockprocessor/blockprocessor.go [hunk 1]
```diff
@@ -54,25 +70,40 @@ func New(
...
+	block, err := bp.buildBlock(coinbaseData, transactions)
+	log.Debugf("BuildBlock end. Took: %s", time.Since(start))
+	return block, err
 }
 
 // ValidateAndInsertBlock validates the given block and, if valid, applies it
 // to the current state
 func (bp *blockProcessor) ValidateAndInsertBlock(block *externalapi.DomainBlock) error {
-	return nil
+	start := time.Now()
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/blockprocessor.go:104]: I suggest maybe have this thing infrastracturized.
a.k.a. Add an infra function 
`func logAndMeasureExecutionTime(funcName string) deferredCallback`
And then you'll call 
`defer logAndMeasureExecutionTime("ValidateAndInsertBlock")()` in the beginning of the function, and be finished with this.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	err = bp.consensusStateManager.AddBlockToVirtual(block.Hash)
+	if err != nil {
+		return err
+	}
+
+	return bp.commitAllChanges()
+}
+
+func (bp *blockProcessor) validateBlock(block *externalapi.DomainBlock) error {
+	// If either header of proof-of-work validation fails, simply
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:32]: typo: of -> or

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+func (bp *blockProcessor) validateHeaderAndProofOfWork(block *externalapi.DomainBlock) error {
+	err := bp.blockValidator.ValidateHeaderInIsolation(block.Hash)
+	if err != nil {
+		return err
+	}
+	err = bp.blockValidator.ValidateHeaderInContext(block.Hash)
+	if err != nil {
+		return err
+	}
+	err = bp.blockValidator.ValidateProofOfWorkAndDifficulty(block.Hash)
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:60]: The order of validations is incorrect. You know the correct one.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	bp.consensusStateStore.Discard()
+	bp.ghostdagDataStore.Discard()
+	bp.multisetStore.Discard()
+	bp.pruningStore.Discard()
+	bp.reachabilityDataStore.Discard()
+	bp.utxoDiffStore.Discard()
+}
+
+func (bp *blockProcessor) commitAllChanges() error {
+	dbTx, err := bp.databaseContext.NewTx()
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:93]: I'd rather have a list of all stores that should be discarded/committed/whatevered, then loop over them in xxxAllChanges

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+	err = bp.blockValidator.ValidateProofOfWorkAndDifficulty(block.Hash)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bp *blockProcessor) validateBody(blockHash *externalapi.DomainHash) error {
+	err := bp.blockValidator.ValidateBodyInIsolation(blockHash)
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:68]: In continuation with the "validation order is incorrect" thingy: just to make sure: this has to happen before the first commitAllChanges.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+	return bp.validateBody(block.Hash)
+}
+
+func (bp *blockProcessor) validateHeaderAndProofOfWork(block *externalapi.DomainBlock) error {
+	err := bp.blockValidator.ValidateHeaderInIsolation(block.Hash)
+	if err != nil {
+		return err
+	}
+	err = bp.blockValidator.ValidateHeaderInContext(block.Hash)
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:56]: In continuation with the "validation order is incorrect" thingy: just to make sure: this has to happen before the first commitAllChanges.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+	// This is to prevent spamming attacks.
+	err := bp.validateHeaderAndProofOfWork(block)
+	if err != nil {
+		return err
+	}
+
+	// It's now safe to store the block in the database
+	bp.blockStatusStore.Stage(block.Hash, model.StatusInvalid)
+	bp.blockStore.Stage(block.Hash, block)
+	err = bp.commitAllChanges()
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:43]: Don't commit here, unless the block failed validation

**Review by svarogg — CHANGES_REQUESTED** [review]
You should handle RuleError in a different way then any other error

File: domain/consensus/processes/coinbasemanager/coinbasemanager.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	return &coinbaseManager{
+		ghostdagDataStore:   ghostdagDataStore,
+		acceptanceDataStore: acceptanceDataStore,
+	}
+}
+
+func (c coinbaseManager) ExpectedCoinbaseTransaction(blockHash *externalapi.DomainHash,
+	coinbaseData *externalapi.DomainCoinbaseData) (*externalapi.DomainTransaction, error) {
+
+	panic("implement me")
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply domain/consensus/processes/coinbasemanager/coinbasemanager.go:27]: Now I did.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go (hunk 6 above)
Comment (stasatdaglabs) [reply domain/consensus/processes/blockprocessor/validateandinsertblock.go]: In-context stuff needs to happen AFTER the first commitAllChanges.

File: domain/consensus/processes/blockprocessor/validateandinsertblock.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	err := bp.validateBlockInIsolationAndProofOfWork(block)
+	if err != nil {
+		return err
+	}
+
+	// If in-context validations fail, discard all changes and store the
+	// block with StatusInvalid.
+	err = bp.validateInContext(block)
+	if err != nil {
+		if errors.Is(err, &ruleerrors.RuleError{}) {
```
Comment (svarogg) [inline domain/consensus/processes/blockprocessor/validateandinsertblock.go:46]: I think you want errors.As here

**Review by svarogg — CHANGES_REQUESTED** [review]

**Pull request merged**